prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala

```diff
@@ -84,41 +81,58 @@ trait Vectors extends Scalars with TimeUnits with Base {
   }
 
   sealed trait Vector extends Expression {
-
     def metricName: Option[String]
     def labelSelection: Seq[LabelMatch]
 
-    def realMetricName: String = {
-      val nameLabelValue = labelSelection.find(_.label == PromMetricLabel).map(_.value)
-      if (nameLabelValue.nonEmpty && metricName.nonEmpty) {
-        throw new IllegalArgumentException("Metric name should not be set twice")
+    // Convert metricName{labels} -> {labels, __name__="metricName"} so it's uniform
+    lazy val mergeNameToLabels: Seq[LabelMatch] = {
+      val nameLabel = labelSelection.find(_.label == PromMetricLabel)
+      if (nameLabel.isEmpty && metricName.isEmpty)
+        throw new IllegalArgumentException("Metric name is not present")
+      if (metricName.nonEmpty) {
+        if (nameLabel.nonEmpty) throw new IllegalArgumentException("Metric name should not be set twice")
+        // metric name specified but no __name__ label.  Add it
+        labelSelection :+ LabelMatch(PromMetricLabel, EqualMatch, metricName.get)
+      } else {
+        labelSelection
       }
-      metricName.orElse(nameLabelValue)
-        .getOrElse(throw new IllegalArgumentException("Metric name is not present"))
     }
 
-    protected def labelMatchesToFilters(labels: Seq[LabelMatch]) =
-      labels.map { labelMatch =>
-        val labelValue = labelMatch.value.replace("\\\\", "\\")
+    def realMetricName: String = mergeNameToLabels.find(_.label == PromMetricLabel).get.value
+
+    // Returns (trimmedMetricName, column) after stripping ::columnName
+    private def extractStripColumn(metricName: String): (String, Option[String]) = {
+      val parts = metricName.split("::", 2)
+      if (parts.size > 1) (parts(0), Some(parts(1))) else (metricName, None)
```

Comment:
Please add this require 

`require(parts(1).nonEmpty, "cannot use empty column name")`